### DISCUSSION TITLE
[NASA] Help understand packet sending

### DISCUSSION THREAD

**Discussion by betaphi** [omerfaruk-aran/esphome_samsung_hvac_bus#43] [General]
Hello @lanwin and thank you very much for all the hard work you have put into the project so far.

I own a Samsung EHS Mono HT Quiet heat pump and as the compressor frequency control among other things is totally messed up I would like to implement such a control by myself. BTW I have plenty of software development and technical background to generally know what I'm doing there ;-)

I have already found out that it is possible to control the compressor frequency via Modbus.
Also, I have written a NASA-Parser, that helps analyse what is going on on the bus as it displays the packets in a more human readable format. See attachment (zipped HTML) [NASA Example.zip](https://github.com/lanwin/esphome_samsung_ac/files/13567708/NASA.Example.zip).

Using this parser I have learned that the indoor unit and outdoor unit are communicating only via broadcast messages. So it seems with NASA there is no "master" or "slave".

I'm asking myself when sending messages via the NASA bus, how can I know when to send a message in order to not interfere with messages from other participants on the bus.

@lanwin did you find any information on that topic in the S-NET implementation? Or is it just sending and hoping optimistically not to collide with other sends?

**Comment by lanwin** [comment]
Hi @betaphi, yes your are right. The entire communication of NASA is based on messages. All units receive that messages and the ones with the desired destination address reacts to that massage. There are some generic addresses like Broadcast etc. Also that entire protocol is pretty generic. Its basically there to interchange variables (MessageNumber its Value).

Currently I have impl a very simple way of sending the messages. I put them in a queue and send them on the receiving thread when there is no more data available. This should reduce collections but costs some time (when there are a lot data available).

Inside the S-NET code there is a way of handling this. Each message have a packetNumber. That is basically and Integer witch is incremented for each messages that is getting sent. 
In example when you change the target temperature you the a message of DataType::Request. Then other side generates a message of data type DataType::Response with the same packetNumber. So basically it should be possible to send a request, wait and check if the reponse comes and if not, resend the request.

I saw code inside the S-NET code but I did not tried it myself yet. I dont now if every request gets a response etc.

Also I noticed that between the messages there are something like filler bytes. The S-NET code calls them permeable, but I have no clue why they are there.

**Comment by betaphi** [comment]
Hello @lanwin, thank you for your response.

The interesting thing about the EHS ASHP is that while I (and others) have recorded the NASA-Stream, it never sends messages with other destinations than just broadcast. I find that pretty notable. Maybe you see different results with the air conditioners.

The preamble you mentioned (and I have noticed also) could in fact be a measure in order to combat collisions given that there is no clear master - slave pattern with NASA.
